Added central, generalized incomplete and regularized beta functions

naveensaigit opened the pull request:
I have added a `betainc` class through which generalized incomplete and generalized regularized incomplete beta functions can be computed. It uses `betainc` from mpmath for calculations in the backend.

#### References to other Issues or PRs
Fixes #20098, Stalled #18727

#### Brief description of what is fixed or changed

#### Other comments
Once this is merged, I plan to work on adding the PERT distribution #16970

#### Release Notes
<!-- BEGIN RELEASE NOTES -->
* functions
  * Added Generalized Incomplete, Generalized Regularized and Central Beta functions
<!-- END RELEASE NOTES -->

naveensaigit committed:
Added central, generalized incomplete and regularized beta functions

naveensaigit committed:
Removed unused variable

oscarbenjamin commented:
Closing and reopening to restart the tests.

oscarbenjamin commented on sympy/functions/special/beta_functions.py:
```diff
@@ -88,6 +97,11 @@ class beta(Function):
+    def __new__(cls, x, y=None):
+        if y == None:
+            y = x
+        return super(cls, cls).__new__(cls, x, y)
```
Function subclasses should define `eval` rather than `__new__`.

naveensaigit replied:
Yes, but how would I make sure that the no. of args are 2 when suppose I try `beta(x)`? Or should I define another class for central beta? It would be a lot of duplicate code

oscarbenjamin replied:
```
def eval(cls, a, b=None):
    if b is None:
         return beta(a, a)
```

naveensaigit replied:
This was the first thing that came to my mind. However, it does not work as `nargs=2`.
``` python
from sympy import *
>>> x=symbols('x')
>>> beta(x,x)
beta(x, x)
>>> beta(x)
Traceback (most recent call last):
  File "/home/naveen/dev/sympy/sympy/core/cache.py", line 72, in wrapper
    retval = cfunc(*args, **kwargs)
  File "/home/naveen/dev/sympy/sympy/core/function.py", line 465, in __new__
    raise TypeError(temp % {
TypeError: beta takes exactly 2 arguments (1 given)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/naveen/dev/sympy/sympy/core/cache.py", line 74, in wrapper
    retval = func(*args, **kwargs)
  File "/home/naveen/dev/sympy/sympy/core/function.py", line 465, in __new__
    raise TypeError(temp % {
TypeError: beta takes exactly 2 arguments (1 given)
```
Setting `nargs=1` doesn't work either since we can't pass in 2 arguments then. This is the reason I overloaded the constructor.

oscarbenjamin replied:
I think `nargs` can be a set like `{1, 2}`

naveensaigit replied:
I just tried it and it raises an error : `ValueError: {1, 2} is not an integer`

oscarbenjamin replied:
This is used by `log`:
```
In [1]: log.nargs
Out[1]: {1, 2}
```

naveensaigit replied:
It works if I don't specify `nargs=2`

oscarbenjamin commented on sympy/functions/special/beta_functions.py:
```diff
@@ -118,3 +132,126 @@ def _eval_conjugate(self):
+    unbranched = True
+
+    def __new__(cls, a, b, x1, x2, regularized=False):
+        return Function.__new__(cls, a, b, x1, x2, int(regularized == True))
```
Storing a boolean flag like this is a bad idea. This should just be two separate classes.

naveensaigit replied:
Yes, I felt that too, but I would have to pass in a boolean flag anyway to the mpmath function. Even if I had another class, say `regularized_betainc`, I would have to pass in the boolean flag 1. Is there a better way to do this?

oscarbenjamin replied:
Passing a boolean flag in just one method is fine. At the moment most methods of the class look like:
```
if self.flag:
    # one thing
else:
    # other thing
```
That's a clear sign that it should be two classes.

If there is code to share between the two classes then it can go in a common superclass.

naveensaigit replied:
Okay, I'll split the code into two different classes.

naveensaigit replied:
If I split the classes and name them `betainc` and `regularized_betainc`, the problem I face with the latter is that I can't use the `mpmath` function which is named `betainc`. Any way to resolve this? @oscarbenjamin

oscarbenjamin replied:
I don't know the implementation of these things so well. Is it not possible to just check the flag before calling into mpmath?

naveensaigit replied:
The `_eval_evalf` of `Function` tries to match the sympy function with mpmath functions and if there exists a function with the same name, it calls it. The code that does this is here - 
https://github.com/sympy/sympy/blob/1e76752f545a2ed44a8af008c76c9ca7dd77e41d/sympy/core/function.py#L536-L561
An easy way to do this would be to define `_imp_` for fallback evaluation, but that would mean that we would have to import mpmath in the function. I can't think of anything else here

oscarbenjamin replied:
Can we not just add this to `MPMATH_TRANSLATIONS`?

naveensaigit replied:
I think this will do the job, thanks!

naveensaigit commented:
@oscarbenjamin Please suggest any more changes that need to be done here

naveensaigit committed:
Added class for regularized beta

naveensaigit commented:
@oscarbenjamin I've made all the requested changes. Let me know if anything else is required.

naveensaigit commented:
@oscarbenjamin Is this good to be merged?

oscarbenjamin commented on sympy/functions/special/beta_functions.py:
```diff
@@ -100,7 +108,9 @@ def fdiff(self, argindex):
-    def eval(cls, x, y):
+    def eval(cls, x, y=None):
+        if y == None:
+            return beta(x, x)
```
Test for `None` with `is`. Also it's more efficient to do:
```
if y is None:
    y = x
```
rather than the recursive call.

naveensaigit replied:
Sure, I'll change that

naveensaigit replied:
If I don't do the recursive call, it outputs an object that looks like a single argument object. Eg- 
``` python
>>> from sympy import beta
>>> x, y = symbols('x y')
>>> beta(x, y)
beta(x, y)
>>> beta(x)
beta(x) #Was previously beta(x, x)
```
So, which is better?

oscarbenjamin replied:
I guess the recursive call is fine then

naveensaigit replied:
Ok

oscarbenjamin commented on sympy/functions/__init__.py:
```diff
@@ -104,7 +104,7 @@
     'elliptic_k', 'elliptic_f', 'elliptic_e', 'elliptic_pi',
 
-    'beta',
+    'beta', 'betainc', 'regularized_betainc',
```
I think it would be better if the name begins with beta somehow. I think Mathematica calls this `BetaRegularised` although I'm not sure if that's exactly the same function.

naveensaigit replied:
Yes, that's the same function. Should I rename it to `beta_regularized`?

oscarbenjamin replied:
Maybe it would be good to survey other languages, Matlab, mpmath, Maple, scipy etc.

naveensaigit replied:
Matlab: `betainc`
mpmath: `betainc`
Maple: Couldn't find one
scipy: `betainc`
Mathematica: `BetaRegularized`
Octave: `betainc`
Maxima: `beta_incomplete_regularized`
So, which name would fit in best?

naveensaigit replied:
Please suggest the name for the function. I think everything else is sorted here.

oscarbenjamin replied:
Looks like `betainc` is the most common name and it's also consistent with both `scipy` and `mpmath` which sympy uses.

Do `scipy` and `mpmath` define the function in exactly the same way as sympy (for all possible complex values)?

naveensaigit replied:
`mpmath` has the same definition of `betainc`. However, the same function is used for both the normal and the regularized incomplete beta function. This is generalized, which means the integral is computed from `x1 to x2`.

The definition of `scipy` is different. They define the regularized incomplete beta as `betainc` and the normal incomplete beta function can be obtained by multiplying with `beta(a, b)`. Further, the integral is always from 
`0 to x`.

oscarbenjamin replied:
I'd suggest to go with betainc then but we will need to check that both mpmath and lambdify/scipy are working correctly.

naveensaigit replied:
But we already have a class `betainc` for the normal incomplete beta function. Should I merge the `regularized_betainc` with `betainc`?

naveensaigit replied:
We can't convert it to scipy because the integral is always from 0 to x, whereas we have x1 to x2 in our case. Conversion is only possible when x1=0.

oscarbenjamin replied:
I don't know these functions well enough to suggest the right name.

It would be good to have names that are consistent with other libraries where possible. Having the same name but for something with a different definition is awkward though.

I think that `betainc_regularized` is better than `regularized_betainc`.

naveensaigit replied:
Yes, I'll change it to `betainc_regularized`.

oscarbenjamin commented on sympy/functions/special/beta_functions.py:
```diff
@@ -118,3 +128,201 @@ def _eval_conjugate(self):
+    unbranched = True
+
+    def __new__(cls, a, b, x1, x2, reg=1):
+        return Function.__new__(cls, a, b, x1, x2, 1)
```
What is `reg` for?

naveensaigit replied:
`reg` is a boolean for regularized. This is the argument we have to pass in to mpmath while evaluating numerically. Is there a way to hide arguments from the user? Like a private argument so to say? If so, we could eliminate it from here

oscarbenjamin replied:
Is `reg` always just `1`? If so then it shouldn't be in the args. Another way should be found to pass `reg=1` to mpmath.

naveensaigit replied:
There is one way, where we define a fallback evaluation function `_imp_` for these classes. See this - 
https://github.com/sympy/sympy/blob/9e33ca3db9137ab00d989992f3f28db4c3db37e0/sympy/core/function.py#L551-L555
Inside `_imp_`, we would have to import `betainc` of mpmath, check if all arguments are non-symbolic and return the answer given by mpmath. For this to work, we should also make sure `betainc` does not match with mpmath in `_eval_evalf`.

I'm not sure how good this is though.

oscarbenjamin replied:
I don't think that the `_imp_` function is suitable because it skips all of the logic below. It's possible that we need to add something in `Function._eval_evalf` to check for a new special method.

naveensaigit replied:
If we're going to do this, it would be easy to remove `reg=1` from args.

naveensaigit replied:
I have added `_eval_mpmath` to handle this case

naveensaigit committed:
Changed name to betainc_regularized

naveensaigit commented:
Please review @oscarbenjamin

oscarbenjamin commented on sympy/functions/special/beta_functions.py:
```diff
@@ -118,3 +128,201 @@ def _eval_conjugate(self):
+    unbranched = True
+
+    def _eval_is_real(self):
+        return all(arg.is_real for arg in self.args)
```
This will give False if the arguments are not known to be real. The `is_real` property is a fuzzy bool that can be `True`, `False`, or `None`. Functions for operating on ordinary bools such as `all` should not be used with fuzzy bool because they do not correctly handle None:
```python
In [13]: x = Symbol('x')

In [14]: betainc(2, 3, 4, x).is_real
Out[14]: False

In [15]: betainc(2, 3, 4, x).subs(x, 5).is_real
Out[15]: True
```
That should be fixed and a test added e.g.:
```
if all(arg.is_real for arg in self.args):
    return True
```
That will give None if any arg gives None or False. Returning None is fine though as it just means "unknown".

Is it necessarily the case that if any of the args is not real the function is not real? If so you could change `all` to fuzzy_and` here which would give `False` is any arg gives `False`.

naveensaigit replied:
> Is it necessarily the case that if any of the args is not real the function is not real? If so you could change all to fuzzy_andhere which would giveFalseis any arg givesFalse`.

When `x1 = x2  = 0`, the integral would always be 0, so this is not necessarily true. I'll just change it to the if condition you suggested above.

naveensaigit replied:
``` python
if all(arg.is_real for arg in self.args):
    return True
```
This will return None when I call `betainc(1, 2, -3*I, 4).is_real`. Is this intended?

oscarbenjamin replied:
Assumptions queries use three-valued logic (TVL) and return True, False or None with None meaning "unknown". It is always logically consistent in TVL to return None even if the result of the query could correctly be True or False. On other hand returning True when the answer should be False or vice-versa is a bug that needs to be fixed. What this means is that we should be cautious and allow None to be returned by default  unless we are absolutely sure that the result is True (or False) given the result of any checks.

So yes when I wrote that I did intend for it to return None in that case. If we want that to return False and it is easy to introduce checks that would return False for this case without *ever* being incorrect then we can add further checking to make this return False. An assumptions query like this should be "cheap" though so only simple checking should be used.

naveensaigit replied:
Sure, thanks for the clarification.

oscarbenjamin commented on sympy/functions/special/beta_functions.py:
```diff
@@ -118,3 +128,201 @@ def _eval_conjugate(self):
+        return Function.__new__(cls, a, b, x1, x2, 1)
+
+    def _eval_is_real(self):
+        return all(arg.is_real for arg in self.args)
```
What if `is_real` gives `None`?

Tests should be added for that case.

oscarbenjamin commented on sympy/functions/special/beta_functions.py:
```diff
@@ -118,3 +128,201 @@ def _eval_conjugate(self):
+
+    def _eval_conjugate(self):
+        a, b, x1, x2, reg = [arg.conjugate() for arg in self.args]
+        return self.func(a, b, x1, x2)
```
This can just be `return self.func(*map(conjugate, self.args))`

oscarbenjamin commented on sympy/functions/special/beta_functions.py:
```diff
@@ -118,3 +128,201 @@ def _eval_conjugate(self):
+        from sympy.integrals.integrals import Integral
+        t = Dummy('t')
+        expr = Integral(t**(a - 1)*(1 - t)**(b - 1), (t, x1, x2))
+        return expr / Integral(t**(a - 1)*(1 - t)**(b - 1), (t, 0, 1))
```
The common expression can be factored out of these two lines.

oscarbenjamin commented:
I get an error from `evalf`:
```python
In [20]: betainc(0, 0, 0, 0).evalf()
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
~/current/sympy/sympy/sympy/core/evalf.py in evalf(x, prec, options)
   1323     try:
-> 1324         rf = evalf_table[x.func]
   1325         r = rf(x, prec, options)

KeyError: betainc

During handling of the above exception, another exception occurred:

TypeError                                 Traceback (most recent call last)
<ipython-input-20-07f71065b9d4> in <module>
----> 1 betainc(0, 0, 0, 0).evalf()

~/current/sympy/sympy/sympy/core/evalf.py in evalf(self, n, subs, maxn, chop, strict, quad, verbose)
   1471             options['quad'] = quad
   1472         try:
-> 1473             result = evalf(self, prec + 4, options)
   1474         except NotImplementedError:
   1475             # Fall back to the ordinary evalf

~/current/sympy/sympy/sympy/core/evalf.py in evalf(x, prec, options)
   1328         if 'subs' in options:
   1329             x = x.subs(evalf_subs(prec, options['subs']))
-> 1330         xe = x._eval_evalf(prec)
   1331         if xe is None:
   1332             raise NotImplementedError

~/current/sympy/sympy/sympy/core/function.py in _eval_evalf(self, prec)
    594             v = func(*args)
    595 
--> 596         return Expr._from_mpmath(v, prec)
    597 
    598     def _eval_derivative(self, s):

~/current/sympy/sympy/sympy/core/expr.py in _from_mpmath(x, prec)
    392             return re + im
    393         else:
--> 394             raise TypeError("expected mpmath number (mpf or mpc)")
    395 
    396     @property

TypeError: expected mpmath number (mpf or mpc)
```

oscarbenjamin commented:
It seems that `lambdify` fails:
```python
In [25]: lambdify((x, y, z, t), betainc(x, y, z, t))(1, 2, 3, 4)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-25-f29af18429d9> in <module>
----> 1 lambdify((x, y, z, t), betainc(x, y, z, t))(1, 2, 3, 4)

<lambdifygenerated-5> in _lambdifygenerated(x, y, z, t)
      1 def _lambdifygenerated(x, y, z, t):
----> 2     return (betainc(x, y, z, t))

TypeError: return arrays must be of ArrayType
```

naveensaigit commented:
`lambdify` fails because it uses `betainc` from scipy by default, which is a 3 argument function. The function that mpmath has and the one I've added here is a generalized version of this function (In scipy, `x1=0` always). It works if `modules=['mpmath']`.
``` python
>>> f=lambdify([a, b, x, y], betainc(a, b, x, y), modules=["mpmath"])
>>> f(1,2,3,4)
mpf('-2.5')
>>> betainc(1,2,3,4).evalf()
-2.50000000000000
```

oscarbenjamin commented:
`Integral(f(x), (x, a, b))` can be `Integral(f(x), (x, 0, b)) - Integral(f(x), (x, 0, a))` so is it possible to compute the generalised function here using the scipy function?

naveensaigit commented:
It is possible, but there might be cases where it would be `oo - oo` and so on, which could complicate things. Even otherwise, how would we map the function to mpmath? If there is a suitable fix for https://github.com/sympy/sympy/pull/20569#discussion_r554430831, we could just make `a=0` a hidden argument. Or else, its better to go with the current API.

oscarbenjamin commented:
> It is possible, but there might be cases where it would be `oo - oo` and so on, which could complicate things.

I suggest that there should be a way to rewrite the expression into a form using e.g. piecewise so that it can handle cases like that. Then we can make sure that lambdify knows to do the rewrite and also knows how to call the scipy function when `x1=0`.

naveensaigit commented:
So you want me to make the interface the same as scipy's `betainc`, with `x1=0`, and `x2=x`?

oscarbenjamin commented:
> So you want me to make the interface the same as scipy's `betainc`, with `x1=0`, and `x2=x`?

What I mean is that we should make it possible to use lambdify in the case where `x1=0`. Then lambdify for the general case could rewrite the function to `betainc(0, a, x, y) - betainc(0, b, x, y)` but wrapped in a Piecewise that can take care of the cases where this would be `oo - oo`.

naveensaigit commented:
> > So you want me to make the interface the same as scipy's `betainc`, with `x1=0`, and `x2=x`?
> 
> What I mean is that we should make it possible to use lambdify in the case where `x1=0`. Then lambdify for the general case could rewrite the function to `betainc(0, a, x, y) - betainc(0, b, x, y)` but wrapped in a Piecewise that can take care of the cases where this would be `oo - oo`.

Oh I see. So in that case, should I add some special case code inside of lambdify to see if an expression contains `betainc(a, b, x1, x2)` and replace it with `betainc(a, b, 0, x2) - betainc(a, b, 0, x1)` or does there exist any property like `_eval_lambdify` to accomplish this?

oscarbenjamin commented:
> does there exist any property like `_eval_lambdify` to accomplish this?

I'm not sure. I imagine that there is something in lambdify already for other cases like this. There can be a method on betainc to rewrite the expression and lambdify could call it somewhere maybe.

naveensaigit commented:
> I'm not sure. I imagine that there is something in lambdify already for other cases like this.

I went through the code for lambdify and there doesn't seem to be any code to handle such cases. Could you please check too?

> There can be a method on betainc to rewrite the expression and lambdify could call it somewhere maybe.

This is straightforward when we pass in only `betainc(a, b, x1, x2)` to lambdify, but how would I accomplish this if the expression passed to lambdify is something like `betainc(a, b, x1, x2)**2`? For eg, I define a property `_eval_lambdify` on betainc, how can I make it cascade?
``` python
>>> from sympy import *
>>> betainc(1,2,3,4)._eval_lambdify()
True
>>> (betainc(1,2,3,4)**2)._eval_lambdify()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'Pow' object has no attribute '_eval_lambdify'
```

oscarbenjamin commented:
> how can I make it cascade?

You can use `replace`.

Also `rewrite` recurses through the expression tree.

oscarbenjamin commented:
> I went through the code for lambdify

Lambdify uses the printers to generate the code:
```console
$ git grep rewrite sympy/printing
sympy/printing/codeprinter.py:    # Functions which are "simple" to rewrite to other functions that
sympy/printing/codeprinter.py:    _rewriteable_functions = {
sympy/printing/codeprinter.py:        elif (expr.func.__name__ in self._rewriteable_functions and
sympy/printing/codeprinter.py:              self._rewriteable_functions[expr.func.__name__] in self.known_functions):
sympy/printing/codeprinter.py:            # Simple rewrite to supported function possible
sympy/printing/codeprinter.py:            return self._print(expr.rewrite(self._rewriteable_functions[expr.func.__name__]))
sympy/printing/mathematica.py:        elif (expr.func.__name__ in self._rewriteable_functions and
sympy/printing/mathematica.py:              self._rewriteable_functions[expr.func.__name__] in self.known_functions):
sympy/printing/mathematica.py:            # Simple rewrite to supported function possible
sympy/printing/mathematica.py:            return self._print(expr.rewrite(self._rewriteable_functions[expr.func.__name__]))
sympy/printing/pretty/pretty.py:        return self._print(ite.rewrite(Piecewise))
sympy/printing/pycode.py:        return self._print(expr.rewrite(Piecewise))
sympy/printing/tests/test_octave.py:def test_automatic_rewrite():
```

oscarbenjamin commented:
> Lambdify uses the printers to generate the code:

It might just be better to defined `_print_betainc` for the relevant code printers though.

naveensaigit commented:
It looks like there are some bugs in mpmath's betainc because of which our approach might not work.
``` python
>>> from mpmath import betainc
>>> betainc(0, 0, 5, 0)
mpc(real='-4.6116860184273879e+18', imag='3.1415926535897931')
>>> betainc(0, 0, 0, 5)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\ProgramData\Anaconda3\lib\site-packages\mpmath\ctx_mp_python.py", line 1035, in f_wrapped
    retval = f(ctx, *args, **kwargs)
  File "C:\ProgramData\Anaconda3\lib\site-packages\mpmath\functions\expintegrals.py", line 116, in betainc
    v = x2**a * ctx.hyp2f1(a, 1-b, a+1, x2) / a
  File "<string>", line 7, in __div__
  File "C:\ProgramData\Anaconda3\lib\site-packages\mpmath\libmp\libmpf.py", line 960, in mpf_div
    raise ZeroDivisionError
ZeroDivisionError
```
Even if the above were to be fixed, there are still some issues.
``` python
>>> from mpmath import betainc as mbeta
>>> from scipy.special import betainc as sbeta
>>> mbeta(0, 0, 0.1, 1.1)
mpc(real='4.5951198501345889', imag='-3.1415926535897931')
>>> sbeta(0, 0, 1.1)
nan
>>> sbeta(0, 0, 0.1)
nan
>>> sbeta(0, 0, 1.1) - sbeta(0, 0, 0.1)
nan
```
There will be inconsistent results because scipy has `x1=0` by default.

oscarbenjamin commented:
Have you tried the master version of mpmath?

Probably best to look for or open issues with mpmath and scipy if those are bugs.

If those are bugs in scipy/mpmath then we should still make sure the code here is correctly calling out to those other libraries so that when they do get fixed sympy will give correct results (although perhaps we should have them emit warnings if we think those functions are potentially unreliable - as long as they always raise or give nan rather than returning an incorrect result then we don't need a warning).

For now the main thing to confirm is that in the cases where they do work we are getting the same answer from all of sympy, mpmath and scipy so that we are sure that the definitions of the functions used match up.

You could check the calculations form sympy's definition by using e.g. series. I've just noticed though that derivatives don't work here:
```python
In [3]: betainc(x, y, z, t).diff(x)
Out[3]: 
∂                      
──(betainc(x, y, z, t))
∂x 
```
Is it possible to define what the derivatives of `betainc` would be?

naveensaigit commented:
> Have you tried the master version of mpmath?

Yes, the issue is not fixed even on the master branch.

> Probably best to look for or open issues with mpmath and scipy if those are bugs.

This not a "bug" in scipy so to speak, as the values are consistent with the definition in scipy. All I meant was that we will sometimes get inconsistent results when if we port this to scipy using lambdify. I am looking in the printer code and will update this soon.

> Is it possible to define what the derivatives of betainc would be?

The derivatives can be added with respect to `z1` and `z2`, but the derivatives with respect to `a` and `b` involves [regularized hypergeometic function](https://functions.wolfram.com/HypergeometricFunctions/HypergeometricPFQRegularized/26/01/01/) which is not yet implemented in sympy. Should I add this as well?

naveensaigit committed:
Fixed evalf problem and added tests for is_real

naveensaigit committed:
Fixed reg by adding _eval_mpmath and added derivatives

naveensaigit commented:
I committed a temporary file by mistake. I have removed it in this commit, but I will rebase later so that this will not be present in the commit history.

naveensaigit committed:
Fixed doctests

naveensaigit committed:
Fixed lambdify

naveensaigit commented:
Please review @oscarbenjamin. I've added the printers for both `betainc` and `betainc_regularized` along with some tests in lambdify. Hopefully, the issue with lambdify is resolved now.

> The derivatives can be added with respect to z1 and z2, but the derivatives with respect to a and b involves regularized hypergeometic function which is not yet implemented in sympy. Should I add this as well?

I've added derivatives with respect to `z1` and `z2`. Let me know if we need the derivatives with respect to `a` and `b` as well.

naveensaigit commented:
ping @oscarbenjamin

oscarbenjamin commented on sympy/printing/latex.py:
```diff
@@ -1157,6 +1157,32 @@ def _print_beta(self, expr, exp=None):
         else:
             return r"\operatorname{B}%s" % tex
 
+    def _print_betainc(self, expr, exp=None):
+        tex = r"\left(%s, %s\right)" % (self._print(expr.args[0]),
+                                        self._print(expr.args[1]))
+
+        if exp is not None:
+            return r"\operatorname{B}_{(%s, %s)}^{%s}%s" % (self._print(expr.args[2]),
+                                                            self._print(expr.args[3]),
+                                                            exp, tex)
+        else:
+            return r"\operatorname{B}_{(%s, %s)}%s" % (self._print(expr.args[2]),
+                                                       self._print(expr.args[3]),
+                                                       tex)
```
I guess this could be like:
```python
def _print_betainc(self, expr, exp=None):
    largs = [self._print(arg) for arg in expr.args]

    tex = r"\left(%s, %s\right)" % (largs[0], largs[1])
    if exp is not None:
        return r"\operatorname{B}_{(%s, %s)}^{%s}%s" % (largs[2], largs[3], exp, tex)
    else:
        return r"\operatorname{B}_{(%s, %s)}%s" % (largs[2], largs[3], tex)
```

oscarbenjamin commented on sympy/printing/pycode.py:
```diff
@@ -995,6 +995,23 @@ def _print_uppergamma(self, expr):
+    def _print_betainc(self, expr):
+        return "({0}({2}, {3}, {5}) - {0}({2}, {3}, {4})) * {1}({2}, {3})".format(
+            self._module_format('scipy.special.betainc'),
+            self._module_format('scipy.special.beta'),
```
Using an f-string might make this a bit clearer

oscarbenjamin replied:
Something like
```python
def _print_betainc(self, expr):
    sp_betainc = self._module_format('scipy.special.betainc')
    sp_beta = self._module_format('scipy.special.beta')
    args = [self._print(arg) for arg in expr.args]
    return f"({sp_betainc}({args[0]}, {args[1]}, {args[3]}) ...
```
Better if we can give reasonable names to the args.

naveensaigit replied:
All the printers seem to use the same type of syntax. Would it be okay to make this different? As for args, we could make it `print_args`

naveensaigit replied:
I've changed it to what you suggested.

oscarbenjamin commented on sympy/functions/special/tests/test_beta_functions.py:
```diff
@@ -21,3 +27,31 @@ def test_beta():
+    a, b, x1, x2 = symbols('a b x1 x2')
+
+    assert isinstance(betainc_regularized(a, b, x1, x2), betainc_regularized)
+    assert isinstance(betainc_regularized(a, b, 0, x1), betainc_regularized)
```
There is `unchanged` for this.

oscarbenjamin commented on sympy/functions/special/beta_functions.py:
```diff
@@ -118,3 +129,227 @@ def _eval_conjugate(self):
+    betainc(conjugate(a), conjugate(b), conjugate(x1), conjugate(x2))
+
+    We can numerically evaluate the Incomplete Beta function
+    to arbitrary precision on the whole complex plane:
```
This should clarify whether or not it means for all arguments on the whole complex plane. The statement isn't quite worded right for a function with more than one parameter

naveensaigit replied:
Well, the statement is true for all arguments on the complex plane. How should I rephrase it?

oscarbenjamin replied:
The function isn't defined on the complex plane. It's defined on something like a product space of 4 complex planes. I think it would be better to be explicit and say "for any complex numbers a, b, x1 and x2".

naveensaigit replied:
Oh, I get it. I will change that

oscarbenjamin commented on sympy/functions/special/beta_functions.py:
```diff
@@ -118,3 +129,227 @@ def _eval_conjugate(self):
+        return Function.__new__(cls, a, b, x1, x2)
+
+    def _eval_mpmath(self, func, args):
+        return func(*args, 1)
```
I wonder if this should actually take a prec argument and try to recursively call evalf

oscarbenjamin replied:
Maybe something like this would be better:
```
_eval_mpmath = getattr(self, '_eval_mpmath', None)
if _eval_mpmath is None:
     func = _get_mpmath_func(self.func.__name__)
else:
     func, args = _eval_mpmath()
```

naveensaigit replied:
The reason I went with this type of syntax is so that we have complete freedom over how the args are rearranged. For example, if I wanted to change the API of `betainc_regularized` from `betainc_regularized(a, b, x, y)` to `betainc_regularized(x, y, a, b)`, all I'd have to do is change the definition in `__new__` and in `_eval_mpmath`
```
def _eval_mpmath(self, func, args):
    return func(args[1], args[3], args[0], args[1], 1)
```
This also solves the problem of the hidden args.

oscarbenjamin replied:
I was thinking that it might be good to for `_eval_mpmath` to be able to refer to a completely different mpmath function. I suppose it can (if it ignores the `func` argument)...

naveensaigit replied:
But would that not mean that we have to import mpmath functions in the sympy codebase apart from them being imported in core? Could you give an example? I'm not sure if I correctly understood what you said

oscarbenjamin replied:
Yes, it would mean that. That's fine though because mpmath is a hard dependency. It is good to keep the imports in one place if possible but we also want to make `evalf` extensible from outside.

naveensaigit replied:
Ok, thanks for clarifying. I'll change this and import the function directly

naveensaigit committed:
Apply suggestions from code review

naveensaigit commented:
Please review @oscarbenjamin

oscarbenjamin commented on sympy/printing/latex.py:
```diff
@@ -1157,6 +1157,24 @@ def _print_beta(self, expr, exp=None):
+        else:
+            return r"\operatorname{B}_{(%s, %s)}%s" % (largs[2], largs[3], tex)
+
+    def _print_betainc_regularized(self, expr, exp=None):
```
These two functions are pretty much identical so should be combined into a helper

oscarbenjamin commented on sympy/core/function.py:
```diff
@@ -566,7 +570,10 @@ def _get_mpmath_func(fname):
+            if _eval_mpmath is None:
+                args = [arg._to_mpmath(prec + 5) for arg in self.args]
+            else:
+                args = [arg._to_mpmath(prec + 5) for arg in args]
```
It would be better to assign `args` in the `if/else` above.

naveensaigit replied:
I thought of the same, but this is a try-except block. I would have to add try-except on the top.
https://github.com/sympy/sympy/blob/38b383368f329fbdf43961c68807f35c9d4964dc/sympy/core/function.py#L572-L576

oscarbenjamin replied:
Can you not just do `args = self.args`?

naveensaigit replied:
Yes, that is better

naveensaigit committed:
Made changes to latex printer and added shell printer

naveensaigit committed:
Fix code quality

naveensaigit commented:
Please review @oscarbenjamin

oscarbenjamin commented on sympy/core/function.py:
```diff
@@ -592,6 +599,8 @@ def bad(m):
         with mpmath.workprec(prec):
             v = func(*args)
+            if v == 0:
+                v = mpmath.mpf(v)
```
Why is this needed?

naveensaigit replied:
When `x1==x2`, mpmath's betainc returns an integer 0 instead of an mpmath object whose value is 0. This causes problems later on in evalf

oscarbenjamin replied:
That sounds like a bug to be reported/fixed upstream in mpmath.

I think we should make a special `betainc_mpmath_fix`  function that wraps mpmath's routine and applies this fix. That way we can acknowledge the bug and work around it without hiding other possible bugs. The `_eval_mpmath` routine can be used to select that instead of mpmath's routine.

naveensaigit replied:
I've added a function `betainc_mpmath_fix` in `beta_functions.py` and opened an issue in mpmath.

naveensaigit committed:
Minor change in evalf

naveensaigit committed:
Fix bug in mpmath betainc and remove check in core/function

oscarbenjamin commented:
Looks good to me. Thanks for this.

Sorry for all the back and forth but I think that in the end this is better as a result. A number of issues have been resolved such as how to extend mpmath that will be useful for other PRs later.

naveensaigit commented:
> Sorry for all the back and forth but I think that in the end this is better as a result. A number of issues have been resolved such as how to extend mpmath that will be useful for other PRs later.

Yes, I think a lot of worthy things were added in this PR that would be useful later on. Thank you for your time reviewing the PR.
